packages/grant-explorer/src/context/QFDonationContext.tsx

```diff
@@ -0,0 +1,282 @@
+    initialQFDonationState.tokenApprovalStatus
+  );
+  const [voteStatus, setVoteStatus] = useState(
+    initialQFDonationState.voteStatus
+  );
+
+  const [indexingStatus, setIndexingStatus] = useState(
+    initialQFDonationState.indexingStatus
+  );
+
+  const providerProps: QFDonationState = {
+    tokenApprovalStatus,
+    setTokenApprovalStatus,
+    voteStatus,
+    setVoteStatus,
+    indexingStatus,
+    setIndexingStatus,
+  };
+
+  return (
+    <QFDonationContext.Provider value={providerProps}>
+      {children}
+    </QFDonationContext.Provider>
+  );
+};
+
+export const QFDonationContext = createContext<QFDonationState>(
+  initialQFDonationState
+);
+
+function resetToInitialState(context: QFDonationState) {
+  const { setTokenApprovalStatus, setVoteStatus, setIndexingStatus } =
+    context;
+
+  setTokenApprovalStatus(
+    initialQFDonationState.tokenApprovalStatus
+  );
+  setVoteStatus(
+    initialQFDonationState.voteStatus
+  );
+  setIndexingStatus(initialQFDonationState.indexingStatus);
+}
+
+async function _submitDonations({
+  signer,
+  context,
+  roundId,
+  donations,
+  donationToken,
+  totalDonation,
+}: SubmitDonationParams) {
+
+  resetToInitialState(context);
+
+  try {
+
+    // Token Approval
+    await approveTokenForDonation(
+      signer,
+      donationToken,
+      totalDonation,
+      context
+    );
+
+    // Invoke Vote
+    const transactionBlockNumber = await vote(
+      signer,
+      roundId,
+      donationToken,
+      donations,
+      context,
+    );
+
+    // Wait for indexing on subgraph
+    await waitForSubgraphToUpdate(signer, transactionBlockNumber, context);
+
+  } catch (error) {
+    datadogLogs.logger.error(`error: _submitDonations - ${error}`);
+    console.error("Error while bulk submitting donations: ", error);
+  }
+}
+
+
+export const useQFDonation = () => {
+  const context = useContext<QFDonationState>(
+    QFDonationContext
+  );
+  if (context === undefined) {
+    throw new Error(
+      "useQFDonation must be used within a QFDonationProvider"
+    );
+  }
+
+  const { signer } = useWallet(); // THIS BREAKS
+
+  const handleSubmitDonations = async ( params: QFDonationParams ) => {
+    return _submitDonations({
+      ...params,
+      signer: signer as Signer,
+      context,
+    });
+  };
+
+  return {
+    submitDonations: handleSubmitDonations,
+    tokenApprovalStatus: context.tokenApprovalStatus,
+    voteStatus: context.voteStatus,
+    indexingStatus: context.indexingStatus,
+  };
+};
+
+async function approveTokenForDonation(
+  signerOrProvider: Signer,
+  token: PayoutToken,
+  amount: number,
+  context: QFDonationState
+): Promise<number> {
+  const { setTokenApprovalStatus } = context;
+
+  try {
+    setTokenApprovalStatus(ProgressStatus.IN_PROGRESS);
+
+    const votingStrategy = ethers.constants.AddressZero; // TODO wire in voting contract address
+
+    // TODO: Skip if token is 0x0 as no approval is needed
```

Comment:
Ideally if token is address 0x0 we don't need to do this step. 
This is a corner case which has to be handled